src/crimson/net/SocketMessenger.cc

```diff
@@ -48,11 +49,16 @@ seastar::future<> SocketMessenger::dispatch(ConnectionRef conn)
   return seastar::keep_doing([=] {
       return conn->read_message()
         .then([=] (MessageRef msg) {
-          if (msg) {
-	    return dispatcher->ms_dispatch(conn, std::move(msg));
-	  } else {
-	    return seastar::now();
-	  }
+          // start dispatch, ignoring exceptions from the application layer
+          auto f = dispatcher->ms_dispatch(conn, std::move(msg))
+            .handle_exception([] (std::exception_ptr eptr) {});
+
+          // chain the future as a continuation of pending_dispatch
+          pending_dispatch = pending_dispatch.then(
+            [f = std::move(f)] () mutable { return std::move(f); });
```

Comment:
I'm wondering if it will cause memory leak if we don't trim ready futures until shutdown...

Reply:
and the way how this PR chains the futures concerns me. how can we discard the fulfilled promises in a timely manner without being blocked by some slow chain segments? probably we can shard the pending dispatch future chains to randomly (or round-robin) to multiple (garbage) bins.

Reply:
thanks for the feedback, i'll rework this part to use `seastar::gate` instead, which only cares about the number of outstanding futures and not their ordering